src/main/java/com/google/devtools/build/remote/client/LogParserUtils.java

```diff
@@ -61,14 +67,104 @@ private FileInputStream openGrpcFileInputStream() throws ParamException, IOExcep
     return new FileInputStream(filename);
   }
 
+  // Returns an ExecuteResponse contained in the operation and null if none.
+  // If the operation contains an error, returns null and populates StringBuilder "error" with
+  // the error message.
+  public static <T extends Message> T getExecuteResponse(
+      Operation o, Class<T> t, StringBuilder error) throws IOException {
+    if (o.getResultCase() == ResultCase.ERROR && o.getError().getCode() != Code.OK.value()) {
+      error.append(o.getError().toString());
+      return null;
+    }
+    if (o.getResultCase() == ResultCase.RESPONSE && o.getDone()) {
+      return o.getResponse().unpack(t);
+    }
+    return null;
+  }
+
+  // Returns a Digest contained in the log entry, and null if none.
+  public static Digest extractDigest(LogEntry entry) {
+    if (!entry.hasDetails()) {
+      return null;
+    }
+    RpcCallDetails details = entry.getDetails();
+    if (details.hasExecute()) {
+      if (details.getExecute().hasRequest()
+          && details.getExecute().getRequest().hasActionDigest()) {
+        return details.getExecute().getRequest().getActionDigest();
+      }
+    }
+    if (details.hasGetActionResult()) {
+      if (details.getGetActionResult().hasRequest()
+          && details.getGetActionResult().getRequest().hasActionDigest()) {
+        return details.getGetActionResult().getRequest().getActionDigest();
+      }
+    }
+    return null;
+  }
+
+  private static List<ExecuteResponse> extractExecuteResponse(List<Operation> operations)
+      throws IOException {
+    ArrayList<ExecuteResponse> result = new ArrayList<>();
+    for (Operation o : operations) {
+      StringBuilder error = new StringBuilder();
+      ExecuteResponse response = LogParserUtils.getExecuteResponse(o, ExecuteResponse.class, error);
```

Comment:
nit: it's the same class, you don't need the LogParserUtils prefix

Reply:
Done.